src/Query/PerBundleQueryAccessHandler.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace Drupal\entity\Query;
+
+use Drupal\Core\Entity\EntityHandlerInterface;
+use Drupal\Core\Entity\EntityTypeBundleInfoInterface;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\user\EntityOwnerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * @see \Drupal\entity\EntityAccessControlHandler
+ * @see \Drupal\entity\EntityPermissionProvider
+ *
+ * @todo Take into account \Drupal\Core\Entity\EntityPublishedInterface
+ */
+class PerBundleQueryAccessHandler implements EntityHandlerInterface, QueryAccessHandlerInterface {
+
+  /**
+   * The entity type.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeInterface
+   */
+  protected $entityType;
+
+  /**
+   * @var \Drupal\Core\Entity\EntityTypeBundleInfoInterface
+   */
+  protected $bundleInfo;
+
+  /**
+   * DefaultQueryAccessHandler constructor.
+   */
+  public function __construct(EntityTypeInterface $entityType, EntityTypeBundleInfoInterface $bundleInfo) {
+    $this->entityType = $entityType;
+    $this->bundleInfo = $bundleInfo;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function createInstance(ContainerInterface $container, EntityTypeInterface $entity_type) {
+    return new static(
+      $entity_type,
+      $container->get('entity_type.bundle.info')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function conditions($operation, AccountInterface $account) {
+    $condition = new Condition('OR');
+
+    // When we would write down \Drupal\entity\EntityAccessControlHandler::checkEntityOwnerPermissions
+    // as boolean logic we would end up with something like this:
+    // (view any ... = 1)
+    //   || (*.uid = account.uid && operation any ... = 1)
+    //   || (*.bundle = bundle1 and (operation any bundle1 == 1))
+    //   || (*.bundle = bundle1 and *.uid = account.uid and (operation any bundle1 == 1))
+    //   || (*.bundle = bundle2 and (operation any bundle2 == 1))
+    
+    // We can though take a couple of shortcuts by checking things in PHP which don't vary per
+    // entity.
+    
+    // No conditions are needed when the user can access all entities anyway.
+    $entity_type_id = $this->entityType->id();
+     if ($account->hasPermission("$operation any {$entity_type_id}")) {
+      return;
+    }
+
+    $bundle_info = $this->bundleInfo->getBundleInfo($entity_type_id);
+    $bundle_key = $this->entityType->getKey('bundle');
+    $has_conditions = FALSE;
+    if ($this->entityType->entityClassImplements(EntityOwnerInterface::class)) {
+      $uid_key = $this->entityType->getKey('uid');
+
+
+      // View own $entity_type permission.
```

Comment:
This class seems to be confused about whether it's just PerBundle, or handling both Per bundle, and EntityType granularities. This is the same dilemma we have in the core permission provider issue. I am not convinced of the need to have two classes, but we could at least have two methods, like we have right now in the module permission provider